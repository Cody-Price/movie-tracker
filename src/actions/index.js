export const loginUser = (userObject) => {
  return {
    type: 'LOGIN_USER',
    userObject
  }
}

export const logout = () => {
  return {
    type: 'LOGOUT_USER'
  }
}

export const addMovies = (movies) => {
  return {
    type: 'ADD_MOVIES',
    movies
  }
}

export const addFavorite = (favorite) => {
  return {
    type: 'ADD_FAVORITE',
    favorite
  }
}