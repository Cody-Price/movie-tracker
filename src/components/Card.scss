@import url('https://fonts.googleapis.com/css?family=Acme');
$acme-font: 'Acme', sans-serif;

.card-text {
  opacity: 0;
  position: absolute;
  z-index:  4;
}

.movie-card {
  position: relative;
  border-radius: 2px;
  box-shadow: 4px 6px 6px rgba(255, 255, 255, 0.6);
  color: white;
  font-family: $acme-font;
  height: 450px;
  margin: 20px;  
  overflow: hidden;
  transition-duration: 0.4s;
  width: 300px;
  &:hover {
    transform: scale(1.1);
    transition-duration: 0.4s;
    cursor: pointer;
    img {
      filter: blur(16px); 
    }
    .card-text {
      width: 80%;
      transition: ease-in .4s;
      transition-delay: .3s;
      height: 80%;
      padding: 3px;
      opacity: 1;
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      align-items: center;
      position: absolute;
      /*left: 0;*/
      margin: 16px;
      font-size: 16px;
      text-align: center;
    }
  }
}

.movie-title {
  font-size: 36px;
  -webkit-text-stroke-width: .1px;
  -webkit-text-stroke-color: black;
}

.poster {
  img {
    width: 100%;
  }
}


.favorite-btn {
  position: absolute;
  z-index: 5;
  left: 200px;
  top: 9px;

  &:disabled {
    background-color: grey;
  }

  &:hover:disabled {
    transform: scale(1);
  }

}