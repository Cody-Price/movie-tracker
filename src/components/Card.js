import React, { Component } from 'react';
import './Card.scss'
import {connect} from 'react-redux';
import { postFavorite, deleteFavorite } from '../API';
import { addFavorite, deleteFavoriteFromStore } from '../actions';

class Card extends Component {
  constructor(props) {
    super(props)
  }
  
  toggleFavorite = () => {
    const {movie, user, favorites} = this.props;
    const matchingFavorite = favorites.find(faveMovie => {
      return faveMovie.movie_id === movie.movie_id
    })
    
    if(!matchingFavorite) {
      postFavorite(movie, user.id);
      this.props.addFavorite(movie);
    } else {
      deleteFavorite(user.id, movie.movie_id);
      this.props.deleteFavoriteFromStore(movie)
    }
  }
  
  render() {

    const faves = this.props.favorites.find(fave => {
      return fave.movie_id === this.props.movie.movie_id
    })
    let favoriteClass = faves ? 'favorite-movie favorite-btn' : 'favorite-btn';

    let loggedIn = this.props.user.name ? false : true;
    const posterPath = `https://image.tmdb.org/t/p/w500/${this.props.movie.poster_path}`
    return (
      <div className="movie-card">
        <button className={favoriteClass}
          disabled={loggedIn} 
          onClick={this.toggleFavorite}
          data-tooltip={loggedIn ? 'Must be signed in to favorite movies' : null}
        >
            Favorite
        </button>
        <article className="card-text">
          <h1 className="movie-title">{this.props.movie.title}</h1>
          <p> {this.props.movie.overview} </p>
        </article>
        <article className='poster'>
          <img src={posterPath} />
        </article>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  user: state.user,
  movies: state.movies,
  favorites: state.favorites
})

const mapDispatchToProps = (dispatch) => ({
  addFavorite: (favorite) => dispatch(addFavorite(favorite)),
  deleteFavoriteFromStore: (favorite) => dispatch(deleteFavoriteFromStore(favorite))
})

export default connect(mapStateToProps, mapDispatchToProps)(Card);