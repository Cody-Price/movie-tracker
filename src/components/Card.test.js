import React from 'react';
import { shallow } from 'enzyme';
import { Card, mapStateToProps, mapDispatchToProps } from './Card'
import { connect } from 'react-redux'
import * as actions from '../actions'

describe('Card', () => {
  let wrapper
  let mockMovie
  let mockFavorites
  let mockUser

  beforeEach(() => {
    mockMovie = {title: 'movie1'}
    mockFavorites = [{title: 'movie1'}, {title: 'movie2'}]
    mockUser = { name: 'Taylor', id: 1 }
    wrapper = shallow(
      <Card
        movie={ mockMovie }
        key={'movie1'}
        favorites= { mockFavorites }
        user={ mockUser }
      />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('mapStateToProps', () => {
    it('should return an object with a user and favorites', () => {
      const mockState = {
        user: mockUser,
        favorites: mockFavorites
      }
      const expected = { user: mockState.user, favorites: mockState.favorites }
      const result = mapStateToProps(mockState)
      expect(result).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the correct params', () => {
      const favorite = mockMovie
      const mockDispatch = jest.fn()
      const actionToDispatch = actions.addFavorite(favorite)
      const result = mapDispatchToProps(mockDispatch)
      result.addFavorite(favorite)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})