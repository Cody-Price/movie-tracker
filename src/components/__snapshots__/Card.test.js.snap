// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    favorites={
      Array [
        Object {
          "title": "movie1",
        },
        Object {
          "title": "movie2",
        },
      ]
    }
    movie={
      Object {
        "title": "movie1",
      }
    }
    user={
      Object {
        "id": 1,
        "name": "Taylor",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="favorite-movie favorite-btn"
          data-tooltip={null}
          disabled={false}
          onClick={[Function]}
        >
          Favorite
        </button>,
        <article
          className="card-text"
        >
          <h1
            className="movie-title"
          >
            movie1
          </h1>
          <p>
             
             
          </p>
        </article>,
        <article
          className="poster"
        >
          <img
            src="https://image.tmdb.org/t/p/w500/undefined"
          />
        </article>,
      ],
      "className": "movie-card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Favorite",
          "className": "favorite-movie favorite-btn",
          "data-tooltip": null,
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Favorite",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="movie-title"
            >
              movie1
            </h1>,
            <p>
               
               
            </p>,
          ],
          "className": "card-text",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "movie1",
              "className": "movie-title",
            },
            "ref": null,
            "rendered": "movie1",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                undefined,
                " ",
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              undefined,
              " ",
            ],
            "type": "p",
          },
        ],
        "type": "article",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            src="https://image.tmdb.org/t/p/w500/undefined"
          />,
          "className": "poster",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "src": "https://image.tmdb.org/t/p/w500/undefined",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "article",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="favorite-movie favorite-btn"
            data-tooltip={null}
            disabled={false}
            onClick={[Function]}
          >
            Favorite
          </button>,
          <article
            className="card-text"
          >
            <h1
              className="movie-title"
            >
              movie1
            </h1>
            <p>
               
               
            </p>
          </article>,
          <article
            className="poster"
          >
            <img
              src="https://image.tmdb.org/t/p/w500/undefined"
            />
          </article>,
        ],
        "className": "movie-card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Favorite",
            "className": "favorite-movie favorite-btn",
            "data-tooltip": null,
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Favorite",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="movie-title"
              >
                movie1
              </h1>,
              <p>
                 
                 
              </p>,
            ],
            "className": "card-text",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "movie1",
                "className": "movie-title",
              },
              "ref": null,
              "rendered": "movie1",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  undefined,
                  " ",
                ],
              },
              "ref": null,
              "rendered": Array [
                " ",
                undefined,
                " ",
              ],
              "type": "p",
            },
          ],
          "type": "article",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              src="https://image.tmdb.org/t/p/w500/undefined"
            />,
            "className": "poster",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "src": "https://image.tmdb.org/t/p/w500/undefined",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "article",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
